docfx_project/articles/bootstrapping-trust/part-01.encryption-everywhere.md

```diff
@@ -0,0 +1,310 @@
+the later parts, consider revisiting this part.
+
+Consider the following diagram of a "mesh" distributed system. This mesh
+could be any type of system such as a mesh of Ziti Routers, or maybe it
+is a system of sensors on an airplane. What they do does not matter.
+What matters is that this system has multiple pieces of software
+connecting amongst themselves. Consider what it means to accomplish this
+using public-key cryptography.
+
+![Image of a simple mesh](./images/mesh.png)
+
+In the diagram above, each system needs:
+
+- a key pair for client & server connections
+- to have the public keys of each system it is connecting to
+
+So what do we need to do? Drop into a CLI and start generating keys on
+each machine. Do that by using these commands:
+
+```
+openssl ecparam -name secp256k1-genkey -param_enc explicit -out private-key.pem
+```
+
+```
+openssl req -new -x509 -key private-key.pem -out server.pem -days 360
+```
+
+Voila - you now have a self-signed certificate (that no one trusts). You
+can repeat this for every piece of software in your mesh network, as it
+initially will be deployed. Preferably, you log into each machine and
+generate the private key there. Moving private keys on and off devices
+is a security risk and frowned upon. For maximum security, you would
+work with a
+[Hardware Security Module (HSM)](https://en.wikipedia.org/wiki/Hardware_security_module)
+such as those available embedded into devices or as a fob.
+
+Then you will need to copy each public certificate to every other
+machine and configure your software so that it trusts that certificate.
+The system will need to repeat this process any time the system adds a
+piece of software. If a machine is compromised, the analogous public
+certificate will need to be untrusted on every node in the mesh.
+
+Wow, this seems like quite a bit of work. Consider what this means when
+adding a node or removing one? Vising each machine and reconfiguring it
+each time is quite a bit of overhead. Let us see how Certificate
+Authorities (CA) can help!
+
+#### CAs & Adding Complexity
+
+A CA enables trust deferral from multiple individual certificates to a
+single certificate. Meaning that instead of trusting each certificate,
+each piece of software will trust the CA. The CA will be used to sign
+every public certificate our software pieces need to use. How does
+"signing" work? We will cover that in parts three and four in-depth. For
+now, the basics will be provided.
+
+Here are the high-level steps:
+
+1. create a CA configuration via OpenSSL CNF files
+2. create the CA
+3. use the CA's public key to sign all of the public certificates
+4. distribute the CA's certificate to every machine
+5. configure the machines certificate store or configure the software
+
+For items one and two, the process can be a bit mystical. There are a
+multitude of options involved in managing a CA. To perform number three,
+you will need to go through the processing of creating CSRs on behalf of
```

Comment:
first time mentioning CSR. again - also detailed in part 3 maybe worth mentioning?

Reply:
added